salt/minion.py

```diff
@@ -1744,6 +1744,9 @@ def _handle_decoded_payload(self, data):
                 self.schedule.functions = self.functions
                 self.schedule.returners = self.returners
 
+        if self.opts["grains_refresh_pre_exec"]:
+            self.opts["grains"] = salt.loader.grains(self.opts, force_refresh=True)
```

Comment:
Wonder if need to consider proxy minion usage and context as in similar code to load the grains in _load_modules, see https://github.com/saltstack/salt/blob/cbb22cfd2bb939864657df99f798443a48bde797/salt/minion.py#L1555-L1567

Reply:
Dunno. Tests passed. Not sure if there are any to check this aspect of proxy operation.

Reply:
The proxy minion should be considered too,  could make a case for some embedded systems with dynamic characteristics which are accessed through a proxy that would benefit from that setting too